### PR TITLE
cmd/snap-update-ns: add test explaining content-layout bugs

### PR DESCRIPTION
**Pull request by zyga**
Snapd is affected by several bugs related to content-layout interaction.
There's a very popular scheme where a snap has two interacting elements
that only work in one special edge case and break regularly:

- A content plug is used to receive content of another snap, typically
  under $SNAP/foo somewhere. Where foo exists and is not created with a
  mimic.
- A layout entry is used to "distribute" $SNAP/foo to /usr/share/foo or
  other locations.

This particular arrangement is used to share essential libraries from a
common "runtime" snap which is not a dedicated base. Several basic tests
explore the actual behavior in the following situations.

- The content is initially connected and a mount namespace is
  constructed. This is the only correct behavior show in this group of
  tests. This is known as test case 1.
- The content is initially connected and a mount namespace is
  constructed. The content is then disconnected. This shows several
  anomalies that are referenced in the code (ordering bug, propagation
  bug). This is known as test case 2.
- The content is initially connected and a mount namespace is
  constructed. The content is then disconnected and re-connected.
  This is known as test case 3.
- The content is initially disconnected and a mount namespace is
  constructed. The content is then connected.
  This is known as test case 4.
- The content interface is initially connected and a mount namespace is
  constructed. The content snap is then refreshed.
  This is known as test case 5.
- The content interface is initially connected and a mount namespace is
  constructed. The application snap is then refreshed.
  This is known as test case 6.

This is related to issues:
- https://warthogs.atlassian.net/browse/SNAPDENG-31644 (flip flop)
- https://warthogs.atlassian.net/browse/SNAPDENG-31645 (propagation)

A support makefile and a set of data files is provided to re-generate
the test data once snapd is improved.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

**Review by bboozzoo — APPROVED** [review]
Thanks

**Comment by ZeyadYasser** [comment]
I found another case that fails consistently a while ago: https://warthogs.atlassian.net/browse/SNAPDENG-21751

The content is initially connected and a mount namespace is
constructed. The snap is a parallel install.

```bash
zeyad@z16:~$ snap install firefox firefox_2
zeyad@z16:~$ sudo nsenter --mount=/run/snapd/ns/firefox.mnt /bin/bash -c "ls /usr/share/libdrm && mount | grep libdrm"
amdgpu.ids # content layout mount is populated
/var/lib/snapd/snaps/gnome-42-2204_176.snap on /usr/share/libdrm type squashfs (ro,relatime,errors=continue,threads=single)
/var/lib/snapd/snaps/core22_1380.snap on /usr/lib/x86_64-linux-gnu/libdrm.so.2.4.0 type squashfs (ro,nodev,relatime,errors=continue,threads=single)
/var/lib/snapd/snaps/gnome-42-2204_176.snap on /usr/share/libdrm type squashfs (ro,relatime,errors=continue,threads=single)
zeyad@z16:~$ sudo nsenter --mount=/run/snapd/ns/firefox_2.mnt /bin/bash -c "ls /usr/share/libdrm && mount | grep libdrm"
tmpfs on /usr/share/libdrm type tmpfs (rw,relatime,mode=755,inode64)
# content layout mount is empty
/var/lib/snapd/snaps/core22_1380.snap on /usr/lib/x86_64-linux-gnu/libdrm.so.2.4.0 type squashfs (ro,nodev,relatime,errors=continue,threads=single)
tmpfs on /usr/share/libdrm type tmpfs (rw,relatime,mode=755,inode64)
```

File: cmd/snap-update-ns/testdata/annotations-1.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ tmpfs /usr/share/secureboot tmpfs x-snapd.synthetic,x-snapd.needed-by=/usr/share/secureboot/potato,mode=0755,uid=0,gid=0 0 0
++# This is the only existing sub-directory of /usr/share/secureboot being
++# re-created by the writable mimic. Note that the name and directory are a lie
++# (or simplification) as this entry exists so that we know it needs to be
++# unmounted. The real source is somewhat more complicated and cannot be
++# expressed with the limited syntax of fstab mount entries.
+ /usr/share/secureboot/updates /usr/share/secureboot/updates none rbind,x-snapd.synthetic,x-snapd.needed-by=/usr/share/secureboot/potato,x-snapd.detach 0 0
++# This is the layout entry replicating the attached content. This is the root
++# of all evil, as the very popular behavior used by nearly all the snaps relies
++# on an implementation dertail that was never envisioned to work this way.
```
Comment (ZeyadYasser) [inline cmd/snap-update-ns/testdata/annotations-1.patch:27]:
```suggestion
+# on an implementation detail that was never envisioned to work this way.
```

**Review by ZeyadYasser — APPROVED** [review]
This is super elegant, Thank you!

I found another case that fails consistently when I started looking into mounts myself a while ago: https://warthogs.atlassian.net/browse/SNAPDENG-21751

The content is initially connected and a mount namespace is
constructed. The snap is a parallel instance.

```bash
zeyad@z16:~$ snap install firefox firefox_2
zeyad@z16:~$ sudo nsenter --mount=/run/snapd/ns/firefox.mnt /bin/bash -c "ls /usr/share/libdrm && mount | grep libdrm"
amdgpu.ids # content layout mount is populated
/var/lib/snapd/snaps/gnome-42-2204_176.snap on /usr/share/libdrm type squashfs (ro,relatime,errors=continue,threads=single)
/var/lib/snapd/snaps/core22_1380.snap on /usr/lib/x86_64-linux-gnu/libdrm.so.2.4.0 type squashfs (ro,nodev,relatime,errors=continue,threads=single)
/var/lib/snapd/snaps/gnome-42-2204_176.snap on /usr/share/libdrm type squashfs (ro,relatime,errors=continue,threads=single)
zeyad@z16:~$ sudo nsenter --mount=/run/snapd/ns/firefox_2.mnt /bin/bash -c "ls /usr/share/libdrm && mount | grep libdrm"
tmpfs on /usr/share/libdrm type tmpfs (rw,relatime,mode=755,inode64)
# content layout mount is empty
/var/lib/snapd/snaps/core22_1380.snap on /usr/lib/x86_64-linux-gnu/libdrm.so.2.4.0 type squashfs (ro,nodev,relatime,errors=continue,threads=single)
tmpfs on /usr/share/libdrm type tmpfs (rw,relatime,mode=755,inode64)
```

File: cmd/snap-update-ns/testdata/annotations-1.patch (hunk 1 above)
Comment (zyga) [reply cmd/snap-update-ns/testdata/annotations-1.patch]: Thanks, fixed.

**Pull request merged**